src/red_black_tree.jl

```diff
@@ -0,0 +1,327 @@
+# it has unique keys
+# leftChild has keys which are less than the node
+# rightChild has keys which are greater than the node
+# color is true if it's a Red Node, else it's false
+mutable struct RBTreeNode{K}
+    color::Bool 
+    data::Union{K, Nothing}
+    leftChild::Union{Nothing, RBTreeNode{K}}
+    rightChild::Union{Nothing, RBTreeNode{K}}
+    parent::Union{Nothing, RBTreeNode{K}}
+
+    RBTreeNode{K}() where K = new{K}(true, nothing, nothing, nothing, nothing)
+    
+    RBTreeNode{K}(d::K) where K = new{K}(true, d, nothing, nothing, nothing)
+end
+
+RBTreeNode() = RBTreeNode{Any}()
+RBTreeNode(d) = RBTreeNode{Any}(d)
+
+function Base.isequal(x::RBTreeNode{K}, y::RBTreeNode{K})where K
+   return (x.color == y.color && x.data == y.data)
+end
+
+Base.:(==)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = isequal(x, y)
+
+Base.:(!=)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = !isequal(x, y)
+
+function create_null_node(K::Type)
+    node = RBTreeNode{K}()
+    node.color = false
+    return node
+end 
+
+mutable struct RBTree{K}
```

Comment:
Should this struct be mutable, or should it just be changed by mutating the root node?

Reply:
I think, it should be `mutable`. Althought the `nil` element is fixed, the `root` keeps on changing.

Reply:
Can we fake that by being able to just mutate the content of the root node?